service/frontend/http_api_server.go

```diff
@@ -0,0 +1,438 @@
+			return nil, fmt.Errorf("failed getting TLS config for HTTP API: %w", err)
+		} else if tlsConfig != nil {
+			listener = tls.NewListener(listener, tlsConfig)
+		}
+	}
+
+	h := &HTTPAPIServer{
+		listener:          listener,
+		logger:            logger,
+		shutdownDrainTime: util.Max(time.Second, serviceConfig.ShutdownDrainDuration()),
+		stopped:           make(chan struct{}),
+	}
+
+	// Build 4 possible marshalers in order based on content type
+	opts := []runtime.ServeMuxOption{
+		runtime.WithMarshalerOption("application/json+pretty+no-payload-shorthand", h.newMarshaler(true, true)),
+		runtime.WithMarshalerOption("application/json+no-payload-shorthand", h.newMarshaler(false, true)),
+		runtime.WithMarshalerOption("application/json+pretty", h.newMarshaler(true, false)),
+		runtime.WithMarshalerOption(runtime.MIMEWildcard, h.newMarshaler(false, false)),
+	}
+
+	// Set Temporal service error handler
+	opts = append(opts, runtime.WithProtoErrorHandler(h.errorHandler))
+
+	// Match headers w/ default
+	h.matchAdditionalHeaders = map[string]bool{}
+	for _, v := range defaultForwardedHeaders {
+		h.matchAdditionalHeaders[v] = true
+	}
+	for _, v := range rpcConfig.HTTPAdditionalForwardedHeaders {
+		h.matchAdditionalHeaders[http.CanonicalHeaderKey(v)] = true
+	}
+	opts = append(opts, runtime.WithIncomingHeaderMatcher(h.incomingHeaderMatcher))
+
+	// Create inline client connection
+	clientConn := newInlineClientConn(
+		map[string]any{"temporal.api.workflowservice.v1.WorkflowService": handler},
+		interceptors,
+	)
+
+	// Create serve mux
+	h.serveMux = runtime.NewServeMux(opts...)
+	err = workflowservice.RegisterWorkflowServiceHandlerClient(
+		context.Background(),
+		h.serveMux,
+		workflowservice.NewWorkflowServiceClient(clientConn),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed registering HTTP API handler: %w", err)
+	}
+	// Set the handler as our function that wraps serve mux
+	h.server.Handler = http.HandlerFunc(h.serveHTTP)
+
+	// Put the remote address on the context
+	h.server.ConnContext = func(ctx context.Context, c net.Conn) context.Context {
+		return context.WithValue(ctx, httpRemoteAddrContextKey{}, c)
+	}
+
+	// We want to set ReadTimeout and WriteTimeout as max idle (and IdleTimeout
+	// defaults to ReadTimeout) to ensure that a connection cannot hang over that
+	// amount of time.
+	h.server.ReadTimeout = serviceConfig.KeepAliveMaxConnectionIdle()
+	h.server.WriteTimeout = serviceConfig.KeepAliveMaxConnectionIdle()
+
+	success = true
+	return h, nil
+}
+
+// Serve serves the HTTP API and does not return until there is a serve error or
+// GracefulStop completes. Upon graceful stop, this will return nil. If an error
+// is returned, the message is clear that it came from the HTTP API server.
+func (h *HTTPAPIServer) Serve() error {
+	err := h.server.Serve(h.listener)
+	// If the error is for close, we have to wait for the shutdown to complete and
+	// we don't consider it an error
+	if err == http.ErrServerClosed {
```

Comment:
Please use errors.Is

Reply:
No, I don't want to check whether it is anywhere in the hierarchy. The docs say that it returns exactly this, I check exactly this on purpose.

Reply:
Was required to change this unfortunately by the linter